cli/src/command/account.rs

```
@@ -110,3 +161,106 @@ impl CommandT for Transfer {
+    use sp_core::serde::{Deserialize, Serialize};
+    use std::collections::HashMap;
+    use std::fs::File;
+    use std::fs::OpenOptions;
+    use std::io::Error as IOError;
+    use std::io::{Seek, SeekFrom};
+
+    type Seed = [u8; 32];
+
+    /// Account
+    ///
+    /// An account has a unique `name` and a unique `seed`,
+    /// from which the account key pair is deterministically derived.
+    #[derive(Debug, Clone)]
+    pub struct Account {
+        pub name: String,
+        pub seed: Seed,
+    }
+
+    /// StoredAccount
+    ///
+    /// The data that is stored in the filesystem relative
+    /// to an [Account]. The [Account::name] is used as the key
+    /// to this value, therefore not included here.
+    #[derive(Serialize, Deserialize)]
+    struct StoredAccount {
+        seed: Seed,
+    }
+
+    #[derive(Debug)]
+    pub enum Error {
+        /// An Account with the given name already exists
+        AlreadyExists(),
+        /// An IOError occurred
+        IOError(IOError),
+        /// A serde json Error occurred
+        JsonError(json::Error),
+    }
+
+    /// Add an account to the storage.
+    ///
+    /// Fails if an [Account] with the given name already exists.
+    /// It can also fail from IO and Serde Json errors.
+    pub fn add(account: Account) -> Result<(), Error> {
+        let mut file = get_file();
+        let mut json: HashMap<String, StoredAccount> =
+            json::from_reader(&file).ok().unwrap_or({ HashMap::new() });
+        if json.contains_key(&account.name.clone()) {
+            return Err(Error::AlreadyExists());
+        }
+
+        json.insert(account.name.clone(), StoredAccount { seed: account.seed });
+        overwrite(&mut file, &json)?;
+        Ok(())
+    }
+
+    /// List all the stored accounts
+    pub fn list() -> Vec<Account> {
+        let file = get_file();
+        let json: HashMap<String, StoredAccount> =
+            json::from_reader(&file).ok().unwrap_or({ HashMap::new() });
```

We’re swallowing errors when the account storage cannot be read. I’d prefer to show fail and show an error to the user.

---

The thing is that showing such errors to users adds no value whatsoever. This should only error if someone but the cli is messing up with the file. If that happens, we just basically start fresh. It seems like fair resilience behaviour.

---

Discussed aside, now the errors are propagated properly 👍